### PR TITLE
os/bluestore/KernelDevice.cc: small fix in buffer flush

### PR DESCRIPTION
**Pull request by tanghaodong25**
should set 'io_since_flush' to 0 after sync operation finish, otherwise data in buffer will not be synced to block device if sync operation get error.

Signed-off-by: haodong tang <haodong.tang@intel.com>

### CODE REVIEW

**Comment by yuyuyu101** [comment]
Actually I don't think it's wrong. If we set flag later and new io comming the gap between fdatasync and "set", we will miss the io needed to be flush

**Comment by tanghaodong25** [comment]
Thanks for your reply! You are right, indeed this situation exits and is it necessary to set 'io_since_flush' to 1 if 'fdatasync' operation get error?

**Comment by majianpeng** [comment]
@yuyuyu101 . Ask other question: The reason why use mutext "// serialize flushers, so that we can avoid weird io_since_flush / races (w/ multipler flushers).". But i think we can remove this only replace 
io_since_flush.read() w/ io_since_flush.exchange(0).   What do you think of it?

**Comment by yuyuyu101** [comment]
for most of flush ops, atomic_t is enough and do thing as expected. For a little racing cases, it may trigger two flush calls which still can be accept. Since flush is a fast path, atomic_t looks better still now.

**Comment by yuyuyu101** [comment]
@tanghaodong25 bluestore.cc now ignore the return value of flush failed, I think you could add error handle in bluestore.cc or just insert assert(0) in KernelDevice::flush

**Comment by liewegas** [comment]
Let's assert(0) for now

**Comment by tanghaodong25** [comment]
a) add 'insert(0)'.
b) use atomic compare and swap to replace mutex lock.

File: src/os/bluestore/KernelDevice.cc [hunk 1]
```diff
@@ -176,15 +176,14 @@ void KernelDevice::close()
 
 int KernelDevice::flush()
 {
-  // serialize flushers, so that we can avoid weird io_since_flush
-  // races (w/ multipler flushers).
-  Mutex::Locker l(flush_lock);
   if (io_since_flush.read() == 0) {
```
Comment (yuyuyu101) [inline src/os/bluestore/KernelDevice.cc]: actually, you need to do atomic compare here.

**Comment by tanghaodong25** [comment]
@yuyuyu101  add more fixes.

```c++
if io_since_flush == 0:
    // ret will be 0, and do nothing
    return;
else
    // io_since_flush will swap to 0
    // ret will be 1
    fdatasync();
```

Thanks.

**Comment by yuyuyu101** [comment]
LGTM

**Pull request merged**